### ISSUE TITLE
Firefox does not use HOME variable

### ISSUE THREAD

**Issue by ydario** [bitwiseworks/mozilla-os2#148]
Firefox stores profile data inside the program directory unless MOZILLA_HOME is set to a different directory.
It should use HOME instead.

Also it does not recognize forward slash in MOZILLA_HOME, only backslashes allowed.

**Comment by dryeo** [comment]
On 02/08/16 12:59 PM, ydario wrote:
> Firefox stores profile data inside the program directory unless
> MOZILLA_HOME is set to a different directory.
> It should use HOME instead.

This would break every install that currently uses the program directory 
to store their profile. Though not recommended, I know of at least a few 
who prefer this setup. It would also break testing or reverting to 
previous versions as well.
Those who prefer using %HOME% can always set MOZILLA_HOME=%HOME% in 
their environments.

**Comment by dmik** [comment]
Well, this can be transparently solved by moving profiles from the program directory to HOME at startup (when no MOZILLA_HOME is set).

**Comment by dryeo** [comment]
On 02/08/16 03:03 PM, Dmitriy Kuminov wrote:
> Well, this can be transparently solved by moving profiles from the
> program directory to HOME at startup (when no MOZILLA_HOME is set).
>

Some PATHS such as the mail location, some extensions and secmod.db 
(passwords?) etc are hardcoded. Moving profiles is not usually trivial.
Profiles that contain mail (SM or TB) can also be quite large, my TB 
profile is currently 1.9GBs while HOME is on a 2GB partition.

**Comment by LewisR** [comment]
Bugzilla is rife with discussions (arguments) over not only where the profiles should be stored, but what to do with existing profiles during upgrades. This is an issue common to all Mozilla projects, but as you point out, Dave, probably of greater impact for SM and TB users, who may have profiles of several GB in size.

Some interesting references:

[Bug 247973 - Profiles should be stored under "Mozilla/Thunderbird" not "Thunderbird" (no vendor set)](https://bugzilla.mozilla.org/show_bug.cgi?id=247973)
[Bug 171561 - Mozilla Firefox creates a profile directory called "Mozilla"](https://bugzilla.mozilla.org/show_bug.cgi?id=171561)
[Bug 393620 - Change SeaMonkey Vendor ID to "Mozilla"](https://bugzilla.mozilla.org/show_bug.cgi?id=393620)

I think the most sensible suggestion I've read was to leave established systems intact, and not move anything or change the location for newly-added profiles in/on such systems.

OS/2 is also an unique case, as historically, we have not had a set place for user data (yes, eCS tried to utilize \HOME\DEFAULT for this, and in Blue Lion, we are setting this to simply \HOME, but the MOZILLA_HOME variable is really the key). Getting back to the beginning of this discussion, though, I would say that we should leave things as they are unless we can tell for certain that there are no existing profiles on the system. If none exist, then it's fine to set/change the default location in the absence of MOZILLA_HOME to %HOME% (or %HOME%\%Vendor%).

Users should be warned not to set/change MOZILLA_HOME after profiles have been created without understanding the manual process of profile migration (see [https://wiki.mozilla.org/SeaMonkey:Moving_Profiles_from_mozilla.org_to_Mozilla](https://wiki.mozilla.org/SeaMonkey:Moving_Profiles_from_mozilla.org_to_Mozilla), as an example).

In short, once created, profile migration can be a major task (pain), and without good reason to do it, should probably be avoided.

**Comment by ydario** [comment]
On other platforms profile data is not in program directory, so users are used to save their data elsewhere.

I agree to preserve existing installations, but for new profiles HOME must be the location.

Backslash must be fixed.

Setting MOZILLA_HOME=%HOME% in config.sys does not work.

**Comment by LewisR** [comment]
I think what Dave meant regarding %HOME% was "the same directory referenced by SET HOME=" and not that he was suggesting that the variable was actually expanded during CONFIG.SYS processing (at least, that's what I meant!). ;-)

If ultimately these apps will be installed via RPM (or any type of installer, this could be set by the installer). Of course, it does no harm to set a reasonable default in the app, and I agree that the slash separator needs to be properly recognized.

**Comment by an64** [comment]
Pleasr do not touch this, or at least  leave option to have profile in  program directory.
I have four versions of FF with own  profiles

**Comment by wztest** [comment]
> Also it does not recognize forward slash in MOZILLA_HOME, only backslashes allowed.
Backslashes in path statements are the default on OS/2. MOZILLA_HOME should not be changed in any way that it will break the current setup. Maybe %Home% should be upgraded to support backslashes as well ;)

**Comment by dmik** [comment]
BTW it looks like the original implementor of `NS_NewNativeLocalFile` on OS/2 misinterpreted the function's purpose. Native here is not to express that native slashes are to be used, it's solely to express that the given path string is given in the native (system) 8-bit charset used to encode file names. No other platform does anything like OS/2 — the only thing the code differs between `NS_NewNativeLocalFile` and `NS_NewLocalFile` is character set encoding (the former takes a 8-bit string and the latter takes a unicode string).

Fixing `NS_NewNativeLocalFile` will surely fix MOZILLA_HOME processing as well as dozens (hundreds) other places because `NS_NewNativeLocalFile` is used everywhere.
